### ISSUE TITLE
Change Spec: Add Support for Assert Equal Value in Addition to Assert Equal Value and Type (Issue #21)

### ISSUE THREAD

**Issue by jimkring** [JKISoftware/Caraya#25]
# Problem
When comparing equal numbers which have different representation (like, I32 and U32), assert "is equal" fails. Although, normal LabVIEW comparison passes. This behavior is not documented, nor is it expected by users; it's confusing.
![assert equal numbers](https://user-images.githubusercontent.com/1911151/39705700-75307f76-520f-11e8-951a-e8a3413b0d03.png)

Note: This was reported in [Issue #21](https://github.com/JKISoftware/Caraya/issues/21)

# Proposed Solution
* Deprecate Existing VI with Undesirable Behavior - The proposed solution is to deprecate the existing VI (remove it from the palette and change it's icon and documentation to reflect its deprecated status) so that users writing new code, will not add the existing VI to their newly created code/tests. Since the existing VI will still exist (even though it's not available on the palette), any existing code will continue to call the existing VI.
* Create a New VI with Desired Behavior - At the same time, we will create a new VI that implements the new, desired behavior of checking for equal value while ignoring type. This new VI will be added to the palette (in the same location the old/deprecated VI was in, previously.)

## Assumptions
* Everyone Has Already Worked Around the Bug (if impacted) - Nobody will have code in production that experiences this error and will rely on us to fix the issue (i.e. people won't let their tests fail because of this unexpected behavior). 
* Some May Rely on Existing Behavior - Some people may actually be relying on this behavior of failing the test if the type is different -- it's a low likelihood, but it's possible if a developer is writing a tool that is operating on variant data (a serialization tool, unit test framework, etc.) and trying to ensure data types match.

## About Deprecation
When we deprecate a VI, we want to achieve two main things:
1. We want users to not add the deprecated VI to any *new* code they create -- they should user the new VI.
2. We want existing users to easily notice existing calls to the deprecated VI, so that they can learn that calling it is no longer recommended. They should then be able to o easily learn about the new/recommended VI, and consider changing their code to use the new one.

## Considerations
* Performance: We don't want the new VI to be slower than the exiting VI -- Caraya was built to be lightweight and fast. To achieve this, we should probably only check the value-only based equality check (which will be slower), after the value+type check (which is fast) fails.

# Alternative Solutions
* Change the behavior of the existing VI so that it ignores type
  * Pros: Easy
  * Cons: Changes behavior of any existing tests and could cause problems for some users who rely on the type checking

**Comment by francois-normandin** [comment]
As I'm implementing this feature, I realize the following:

1. We should **not** deprecate the "Assert Equal_Variant" but simply document it better and change its icon to make it obvious that this is a **Variant** comparison (and therefore it should match exactly both value and type to assert True). For reasons stated by @jimkring, we cannot touch its code so we can preserve backward compatibility, but I think it should also be left in the palette with an icon and description change. (Add "Variant" to the icon perhaps?)

2. The "Assert Equal_Value&Type" method will have the same Variant inputs as the "Assert Value_Variant" method already existing: it would need to be polymorphic to differentiate itself from "Assert Value_Variant" in any way, because otherwise, it is exactly the same code under the hood.
So it brings the question: 
- Should we even have this node?
- Shouldn't we simply document the "Assert Equal_Variant" and make it clear that it is a "Value & Type" assertion? 

3. (Depending on point 2 above) To maintain the current naming convention in Caraya, all assertions that accept Variants as inputs are named with the suffix "_Variant.vi". The name of the new proposed method will be "Assert Equal_Value&Type_Variant"... and it will simply wrap the "Assert Equal_Variant" method: fast, simple & obvious self-documenting name for developers to pick from the palette.

4. The new "Assert Equal_Value_Variant" method will work as describe in the spec change above. First calling the Equal_Variant (fast) and then a slower function if it failed to verify that the value is equal.

5. Finally, we need to mirror the **NotEqual_Variant** method with a NotEqual_Value_Variant (to show it asserts false only if value is different, irrespective of type difference. 

### In a nutshell,
I propose we do not implement this spec change and come up with a simpler task: 
- document better the "Assert Equal_Variant" method
- create new methods: "Assert EqualValue_Variant" & "Assert NotEqualValue_Variant"

**Comment by jimkring** [comment]
@francois-normandin It feels to me that the "_Variant" suffix is being (or should be) used to describe the *data type of the argument* to the assertion function, and we should not use this "_Variant" suffix to imply *how* that argument will be evaluated by the assertion.  My thinking would be to use the following names (to be consistent with the current conventions) for two *new* VIs that are added to the palette:

- Assert Equal Value_Variant.vi
- Assert Equal Value and Type_Variant.vi

And, I think we should deprecate "Assert Equal_Variant.vi" since it's name does not correctly/fully convey its behavior -- that's the confusing part that we're rectifying with the new VIs and deprecation process. It forces the user to take a look at why they are calling the old VI that checks type and value and then make a decision about which behavior they want, moving forward.

And, I'm fine if we wish to make the new implementation of "Assert Equal_Variant" simply a call to the new "Assert Equal Value and Type_Variant.vi" VI :-)

Regarding Assert NotEqual, yes we should follow the same path we choose for the changes to Assert Equal function.

**Comment by francois-normandin** [comment]
@jimkring 
Agreed then. Will do.

**Comment by francois-normandin** [comment]
Spec change implemented in release 0.6.0
